libpam/pam_handlers.c

```diff
@@ -571,94 +583,210 @@ int _pam_init_handlers(pam_handle_t *pamh)
+				      char *start, char *end)
+{
+    if (buffer->assembled != NULL || (*start != '\0' && *start != '\n'))
+	return _pam_buffer_add(buffer, start, end);
+    return 0;
+}
+
+static void _pam_buffer_clear(struct line_buffer *buffer)
+{
+    _pam_drop(buffer->assembled);
+    _pam_drop(buffer->chunk);
+    buffer->chunk_size = 0;
+    buffer->len = 0;
+    buffer->size = 0;
+}
+
+static void _pam_buffer_init(struct line_buffer *buffer)
+{
+    buffer->assembled = NULL;
+    buffer->chunk = NULL;
+    _pam_buffer_clear(buffer);
+}
+
+static void _pam_buffer_purge(struct line_buffer *buffer)
+{
+    _pam_drop(buffer->chunk);
+    buffer->chunk_size = 0;
+}
+
+static void _pam_buffer_shift(struct line_buffer *buffer)
+{
+    if (buffer->assembled == NULL)
+	return;
+
+    _pam_buffer_purge(buffer);
+    buffer->chunk = buffer->assembled;
+    buffer->chunk_size = buffer->size;
+
+    buffer->assembled = NULL;
+    buffer->size = 0;
+    buffer->len = 0;
+}
+
+static inline int _pam_buffer_valid(struct line_buffer *buffer)
+{
+    return buffer->assembled != NULL && *buffer->assembled != '\0';
+}
+
+static inline char *_pam_str_trim(char *str)
+{
+	/* skip leading spaces */
+	str += strspn(str, " \t");
+	/*
+	 * we are only interested in characters before the first '#'
+	 * character
+	 */
+	str[strcspn(str, "#")] = '\0';
+
+	return str;
+}
+
+static inline int _pam_str_unescnl(char *start, char **end)
+{
+    int esc = 0;
+    char *p = *end;
+
+    /*
+     * Check for backslash by scanning back from the end of
+     * the entered line, the '\n' should be included since
+     * normally a line is terminated with this character.
+     */
+    while (p > start && ((*--p == ' ') || (*p == '\t') || (*p == '\n')))
+	;
+    if (*p == '\\') {
+	*p++ = ' ';         /* replace backslash with ' ' */
+	*p = '\0';          /* truncate the line here */
+	*end = p;
+	esc = 1;
+    }
+
+    return esc;
+}
+
+static inline int _pam_str_prepare(char *line, ssize_t len,
+				   char **start, char **end)
+{
+    int esc;
+
+    *start = line;
+    *end = line + len;
+
+    esc = _pam_str_unescnl(*start, end);
+    *start = _pam_str_trim(*start);
+
+    return esc;
+}
+
 /*
  * This is where we read a line of the PAM config file. The line may be
  * preceded by lines of comments and also extended with "\\\n"
  */
 
-static int _pam_assemble_line(FILE *f, char *buffer, int buf_len)
+static int _pam_assemble_line(FILE *f, struct line_buffer *buffer)
 {
-    char *p = buffer;
-    char *endp = buffer + buf_len;
-    char *s, *os;
-    int used = 0;
+    int added = 0;
 
     /* loop broken with a 'break' when a non-'\\n' ended line is read */
 
     D(("called."));
+
+    _pam_buffer_shift(buffer);
+
     for (;;) {
-	if (p >= endp - 1) {
-	    /* Overflow */
-	    D(("overflow"));
-	    return -1;
-	}
-	if (fgets(p, endp - p, f) == NULL) {
-	    if (used) {
+	char *start, *end;
+	ssize_t n;
+	int strcont;
```

Comment:
I don't understand what this variable does.

Reply:
It indicates if we have to continue adding file lines to the current configuration line stored in buffer (1) or if we have actually encountered the EOL (\n) which also completes the configuration line.

Reply:
Perhaps `multiline` would be a better name?

Reply:
I've changed the 0/1 logic so we have a boolean which indicates if the current line has an "end of line" marker (i.e. \n) in it or not. This way, the new variable name `eol` indicates just that. If it is true, we have reached the end. Otherwise we have to continue.

I gave your `multiline` approach a thought and noticed that if we reach the end of a multiline, it would be 0 (which feels a tiny bit off, given that we definitely had a multiline in the other iterations). This made me come up with `eol` instead. What do you think?

Reply:
My first idea was to call it `eol`m but that would have reversed your logic and I didn't want to do that. So, now I'm very happy :smile: